lib/drivers/docker/index.js

```diff
@@ -0,0 +1,294 @@
+    instance.currentImage = toLaunch;
+    delete instance.next;
+    instance.current.on('request', function(req) {
+      debug('bubbling up request from instance %j:', id, req);
+      self.emit('request', id, req);
+    });
+    instance.current.on('error', function(err) {
+      console.error('error from docker child %j:', id, err);
+    });
+    instance.current.on('starting', function() {
+      instance.ports = [];
+      if (previous) {
+        previous.kill('SIGKILL', function() {
+          previous.destroy();
+        });
+      }
+    });
+    instance.current.on('exit', function(err) {
+      if (isError(err) && err.statusCode === 404) {
+        self.startInstance(id, function() {
+          console.log('restarted instance after container disappeared (exit)');
+        });
+      }
+    });
+    if (cb) {
+      instance.current.on('connected', cb);
+    }
+  });
+};
+
+DockerDriver.prototype._onRequest = function(instanceId, req) {
+  debug('DockerDriver._onRequest(%j)', instanceId, req);
+  var instance = this._instance(instanceId);
+  switch (req.cmd) {
+    case 'listening':
+      // emit a listening event each time a _new_ port is listened on
+      if (instance.ports.indexOf(req.address.port) < 0) {
+        instance.ports.push(req.address.port);
+        this.emit('listening', instanceId, req.address);
+      }
+  }
+};
+
+DockerDriver.prototype.stopInstance = function(id, style, cb) {
+  debug('DockerDriver.stopInstance(%j, %j, %s)', id, style, cb.name);
+  var instance = this._instance(id);
+  instance.shouldRestart = false;
+  switch (style) {
+    case 'soft':
+      return this.requestOfInstance(id, {cmd: 'stop'}, cb);
+    case 'hard':
+    default:
+      if (instance.current) {
+        return instance.current.kill('SIGTERM', reportExit);
+      } else {
+        return cb(Error('No such instance: ' + id));
+      }
+  }
+  function reportExit(err, res) {
+    debug('stopped container for instance %j:', id, err, res);
+    cb(err, res);
+  }
+};
+
+DockerDriver.prototype._onContainerStopped = function(id, app) {
+  console.error('container stream stopped for some reason!', id, app);
+};
+
+DockerDriver.prototype.dumpInstanceLog = function(id) {
+  debug('DockerDriver.dumpInstanceLog(%j)', id);
+  var instance = this._instance(id);
+  if (!instance.current) {
+    debug('no instance to dump logs for: %j', id);
+    return null;
+  }
+  var logDump = instance.log.duplicate().toString();
+  instance.log.consume(logDump.length);
+  return logDump;
+};
+
+DockerDriver.prototype.updateInstanceEnv = function(id, env, cb) {
+  debug('DockerDriver.updateInstanceEnv(%j)', id);
+  var self = this;
+  var current = this._instance(id).current;
+  if (!current) {
+    debug('no instance to update env of: %j', id);
+    return setImmediate(cb);
+  }
+  // Environment variables can only really have string values, so make sure we
+  // are consistent with that so our comparisons are reliable.
+  env = _.mapValues(env, String);
```

Comment:
should be done as close to mesh models as possible, service-manager, or perhaps even in mesh models.